### ISSUE TITLE
RFC: Get the EC curve name used in key agreement

### ISSUE THREAD

**Issue by johnma14** [haproxy/haproxy#2165]
### Your Feature Request

There are a number of sample fetch methods to get the symmetric cipher key size, name of used ciphers, algorithms used to generate key of certificates, client-hello supported elliptic curve point formats and elliptic curves. This feature request is to 
create a new sample fetch method to get the EC curve name used in the key agreement to enable better observability.

### What are you trying to do?

With a new sample fetch to get the EC curve name, we should be able to use it to get the elliptic curve name used in the key exchange phase of the TLS handshake.

### Output of `haproxy -vv`

```plain
HAProxy version 2.7.6-4dadaaa 2023/03/28 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.6.html
Running on: Linux 5.15.0-71-generic #78-Ubuntu SMP Tue Apr 18 09:00:29 UTC 2023 x86_64
Build options :
  TARGET  = generic
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_SYSTEMD=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES -ACCEPT4 -BACKTRACE -CLOSEFROM -CPU_AFFINITY -CRYPT_H -DEVICEATLAS -DL -ENGINE -EPOLL -EVPORTS -GETADDRINFO -KQUEUE -LIBCRYPT -LINUX_SPLICE -LINUX_TPROXY -LUA -MEMORY_PROFILING -NETFILTER -NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL -PRCTL -PROCCTL -PROMEX -PTHREAD_EMULATION -QUIC -RT -SHM_OPEN +SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD -TFO -THREAD -THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 3.1.0 14 Mar 2023
Running on OpenSSL version : OpenSSL 3.0.2 15 Mar 2022 (VERSIONS DIFFER!)
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Support for malloc_trim() is enabled.
Built without multi-threading support (USE_THREAD not set).
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): no
Built with gcc compiler version 11.3.0

Available polling systems :
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 2 (2 usable), will use poll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by johnma14** [comment]
I can help contribute to this feature if this sounds reasonable to add. Thanks.

**Comment by wlallemand** [comment]
This seems like a useful feature to me, but I must confess I have no idea how to get the curve name from the SSL session. boringssl has `SSL_get_curve_id()` and  `SSL_get_curve_name()` which allow it to do it easily using the SSL * of the current session. But I don't see anything like this in OpenSSL.

I just checked in openssl code and from what I'm seeing, you could use `SSL_get_negotiated_group()` since 3.0 and you could use `OSSL_EC_curve_nid2name()` to convert them to curve names.

**Comment by wlallemand** [comment]
I'm seeing this PR https://github.com/openssl/openssl/pull/20866 which seems to implement SSL_get_curve_name()

**Comment by wlallemand** [comment]
Feel free to propose a patch if you want to try to implement this feature, you could take a look at `smp_fetch_ssl_fc_cipher()` in `src/ssl_sample.c` which is a good example. Your fetch could be named `ssl_fc_curve`.

**Comment by johnma14** [comment]
Thank you @wlallemand , appreciate your input and feedback. I will send in a patch soon.

**Comment by wlallemand** [comment]
This was implemented in 00b7b49. closing.
